actor/src/main/scala/org/apache/pekko/pattern/RetrySupport.scala

```diff
@@ -172,30 +175,34 @@ object RetrySupport extends RetrySupport {
       }
     }
 
-    require(maxAttempts >= 0, "Parameter maxAttempts must >= 0.")
+    def doRetry(nextAttempt: Int): Future[T] = delayFunction(nextAttempt) match {
+      case Some(delay) =>
+        if (delay.length < 1)
+          retry(attempt, maxAttempts, delayFunction, nextAttempt, shouldRetry)
+        else
+          after(delay, scheduler) {
+            retry(attempt, maxAttempts, delayFunction, nextAttempt, shouldRetry)
+          }
+      case None =>
+        retry(attempt, maxAttempts, delayFunction, nextAttempt, shouldRetry)
+      case null =>
+        Future.failed(new IllegalArgumentException("The delayFunction of retry should not return null."))
+    }
+
     require(attempt != null, "Parameter attempt should not be null.")
+    require(maxAttempts >= 0, "Parameter maxAttempts must >= 0.")
+    require(delayFunction != null, "Parameter delayFunction should not be null.")
+    require(attempted >= 0, "Parameter attempted must >= 0.")
+
     if (maxAttempts - attempted > 0) {
       val result = tryAttempt()
       if (result eq null)
         result
       else {
-        val nextAttempt = attempted + 1
-        result.recoverWith {
-          case NonFatal(_) =>
-            delayFunction(nextAttempt) match {
-              case Some(delay) =>
-                if (delay.length < 1)
-                  retry(attempt, maxAttempts, delayFunction, nextAttempt)
-                else
-                  after(delay, scheduler) {
-                    retry(attempt, maxAttempts, delayFunction, nextAttempt)
-                  }
-              case None =>
-                retry(attempt, maxAttempts, delayFunction, nextAttempt)
-              case null =>
-                Future.failed(new IllegalArgumentException("The delayFunction of retry should not return null."))
-            }
-
+        result.transformWith {
+          case Success(value) if shouldRetry(value, null)                        => doRetry(attempted + 1)
+          case Failure(e) if NonFatal(e) && shouldRetry(null.asInstanceOf[T], e) => doRetry(attempted + 1)
+          case _                                                                 => result
```

Comment:
Should I change the code to `isInstanceOf` and `asInstanceOf` for performance?

Still need more time to finish this PR.